lua/orgmode/org/hyperlinks.lua

```diff
@@ -1,61 +1,27 @@
 local Files = require('orgmode.parser.files')
 local utils = require('orgmode.utils')
+local fs = require('orgmode.utils.fs')
 local Hyperlinks = {}
 
-local function get_file_from_context(ctx)
-  local filepath = (ctx.hyperlinks and ctx.hyperlinks.filepath)
-  if not filepath then
-    return Files.get_current_file()
-  end
-  local canonical = vim.loop.fs_realpath(filepath)
-  if not canonical then
-    return Files.get_current_file()
-  end
-
-  return Files.get(canonical)
-end
-
-local function update_hyperlink_ctx(ctx)
-  if not ctx.line then
-    return
-  end
-
-  -- TODO: Support text search, see here [https://orgmode.org/manual/External-Links.html]
-  local hyperlinks_ctx = {
-    filepath = false,
-    headline = false,
-    custom_id = false,
-  }
-
-  local file_match = ctx.line:match('file:(.-)::')
-  if file_match then
-    file_match = Hyperlinks.get_file_real_path(file_match)
-  end
-
-  if file_match then
-    file_match = vim.loop.fs_realpath(file_match)
-  end
-
-  if file_match and Files.get(file_match) then
-    hyperlinks_ctx.filepath = Files.get(file_match).filename
-    hyperlinks_ctx.headline = ctx.line:match('file:.-::(%*.-)$')
-
-    if not hyperlinks_ctx.headline then
-      hyperlinks_ctx.custom_id = ctx.line:match('file:.-::(#.-)$')
-    end
-
-    ctx.base = hyperlinks_ctx.headline or hyperlinks_ctx.custom_id or ctx.base
-  end
-
-  ctx.hyperlinks = hyperlinks_ctx
+---@param url Url
+local function get_file_from_url(url)
+  local file_path = url:get_filepath()
+  local canonical_path = file_path and fs.get_real_path(file_path)
+  return canonical_path and Files.get(canonical_path) or Files.get_current_file()
 end
 
-function Hyperlinks.find_by_filepath(ctx)
+---@param url Url
+---@return string[]
+function Hyperlinks.find_by_filepath(url)
   local filenames = Files.filenames()
-  local file_base = ctx.base:gsub('^file:', '')
+  local file_base = url:get_filepath()
+  if not file_base then
+    return {}
+  end
+  --TODO integrate with orgmode.utils.fs or orgmode.objects.url
   local file_base_no_start_path = file_base:gsub('^%./', '') .. ''
   local is_relative_path = file_base:match('^%./')
-  local current_file_directory = vim.fn.fnamemodify(utils.current_file_path(), ':p:h')
+  local current_file_directory = fs.get_current_file_dir()
   local valid_filenames = {}
   for _, f in ipairs(filenames) do
     if is_relative_path then
```

Comment:
@kristijanhusak in the next line the behavior for relative paths is specified in case of agenda files.
It is very simple: When you are in a directory with agenda files directly in it, you will see them in the completion. Otherwise they are explicitly ignored.
A simple fix would be to replace the `[^/]*` with a `.*` match org files, which are deeper nested from the current path.
To find the relative path to all agenda files, no matter where we are relative to them (e.g. a sibling file tree) would require to implement more complex logic.
So what do you think, which behavior should it be for the merge?